Sources/NetworkProtection/PacketTunnelProvider.swift

```diff
@@ -1215,6 +1219,60 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
         }
     }
 
+    lazy var failureRecoveryHandler: FailureRecoveryHandling = FailureRecoveryHandler(deviceManager: deviceManager)
+
+    private func startServerFailureMonitor() async {
+        if await serverFailureMonitor.isStarted {
+            await serverFailureMonitor.stop()
+        }
+
+        if isSubscriptionEnabled, await isEntitlementInvalid() {
+            return
+        }
+
+        await serverFailureMonitor.start { [weak self] result in
+            guard let self else {
+                return
+            }
+
+            guard case .failureDetected = result else {
+                return
+            }
+
+            Task {
+                guard let server = await self.lastSelectedServer else {
+                    os_log("🟢 No last server info found for recovery", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+                    return
+                }
+                await self.startReasserting()
+                await self.serverFailureMonitor.stop()
+
+                let recoveryResult: (tunnelConfig: TunnelConfiguration, server: NetworkProtectionServer)
+                do {
+                    recoveryResult = try await self.failureRecoveryHandler.attemptRecovery(
+                        to: server,
+                        includedRoutes: self.includedRoutes ?? [],
+                        excludedRoutes: self.settings.excludedRanges,
+                        isKillSwitchEnabled: self.isKillSwitchEnabled
+                    )
+                    await self.handleFailureRecovery(tunnelConfig: recoveryResult.tunnelConfig, server: recoveryResult.server)
+                } catch {
+                    // Pixel here sent as part of https://app.asana.com/0/0/1206939413299475/f
+                    os_log("🟢 Failure recovery error: %{public}@", log: .networkProtectionServerFailureRecoveryLog, type: .error, String(reflecting: error))
+                }
+            }
+        }
+    }
+
+    @MainActor
+    private func handleFailureRecovery(tunnelConfig: TunnelConfiguration, server: NetworkProtectionServer) {
+        self.lastSelectedServer = server
+        Task {
+            try await self.updateAdapterConfig(tunnelConfiguration: tunnelConfig, reassert: true)
```

Comment:
What happens here if the adapter fails to update and throws an error?

Should we add handling?  Should we fire pixels?

Reply:
Yes! https://github.com/duckduckgo/BrowserServicesKit/pull/791/files